rfd/0196-introducing-a-component-system.md

```diff
@@ -0,0 +1,518 @@
+
+### Easy styling
+
+As mentioned above, Chakra adds the ability to create recipes, which allow the definition of reusable components to
+stay in
+one place, and be reused throughout the codebase. Recipes also make defining dark mode variants much easier.
+
+```tsx
+export const styledButton = defineRecipe({
+  base: {
+    border: '1px solid',
+    borderColor: {
+      _dark: 'alpha.100',
+      value: 'alpha.200',
+    },
+    borderRadius: 'md',
+  },
+})
+```
+
+### CSS variables everywhere
+
+Chakra UI exports all defined colors, fonts, spacing, radii, etc. as CSS variables, which means that they can be
+referred to
+when not using Chakra. CSS variables are faster for the browser to parse over grabbing values from the theme prop in a styled component.
+
+If a developer still wishes to use `styled.x`, they can refer to the CSS variables directly, like so:
+
+```tsx
+const StyledBox = styled.div`
+  padding: var(--teleport-space-1) var(--teleport-space-1);
+  background: var(--teleport-colors-levels-popout);
+  border: 1px solid var(--teleport-colors-alpha-200);
+  border-radius: var(--teleport-radii-2);
+`;
+```
+
+However, this is not recommended, as there is no validation of the CSS variables.
+
+### Supports code generation for types
+
+All variants, sizes, etc. are generated as types, which means that the developer experience is much better when using
+components defined from Chakra UI, or built on top of it. Theming props are auto generated for custom recipes, so the
+different variant props can be type-checked and autocompleted during development.
+
+## Reasonable approach to adoption
+
+### First step
+
+After a small proof of concept, full stack resources would be needed to style some of the existing Chakra components to
+Teleport's theme. This can also be a great opportunity for everyone to try it out and get to grips with Chakra.
+
+### Code structure
+
+Converting all the previous codebase to Chakra UI would be a huge task and could introduce regressions. Instead, we
+could
+start by deprecating the previous design system, and any code that uses it is kept in a separate package.
+
+This would also give the advantage of being able to enable stricter ESLint rules, for all new code going forward using
+the new design system. This can help reduce bugs by preventing common mistakes, and also help with maintaining a
+more consistent code style.
+
+```
+# current directory structure
+
+web/
+  packages/
+    design/
+    teleport/
+    teleterm/
+```
+
+Possible new directory structure
+
+```
+web/
+  packages/
+    legacy/ # current ESLint rules, not very strict
+      design/
+        src/
+      teleport/
+        src/
+      teleterm/
+        src/
+    design/
+      src/ # strict ESLint, TypeScript rules
+    teleport/
+      src/ # strict ESLint, TypeScript rules
+    teleterm/
+      src/ # strict ESLint, TypeScript rules
+```
```

Comment:
Just to note, I tested this out, and it didn't work how I expected - you can't have TypeScript only check a certain directory in strict mode whilst it's importing code from places that aren't strict checked. I'm testing out a few other solutions and will update this RFD if it works

(I'm testing out a way that avoids moving the code around which will make backporting even easier)

Reply:
Could [project references](https://www.typescriptlang.org/docs/handbook/project-references.html) help with this?